pkg/operator/reports.go

```diff
@@ -734,6 +742,73 @@ func (op *Reporting) runReport(logger log.FieldLogger, report *metering.Report)
 	return nil
 }
 
+func deleteReport(report *metering.Report, op *Reporting) {
+	if reportNotUsedAsInput(report, op) {
+		reports, err := op.reportLister.Reports(report.Namespace).List(labels.Everything())
+		if err != nil {
+			op.logger.WithError(err).Errorf("on deleting report: %s, for expiration, failed to list reports for namespace: %s",
+				report.Name, report.Namespace)
+		}
+		for _, report := range reports {
+			op.deleteReport(report)
+		}
+		op.logger.Infof("deleted report after retention period: %s, past reportingEnd: %s",
+			report.Spec.Expiration, report.Spec.ReportingEnd)
+	}
```

Comment:
I need to double-check how we're passing the `@report` object to this function, but I worry doing something like this will invalidate the cache and we should avoid deleting any object we're currently processing in this part of the reconciliation process.

Reply:
And just to clarify, marking the object as a deletion candidate is probably fine, with the intention that a higher-level function, like this object's updateFunc will take care of managing the deletion of that resource, but deleting the object outright may cause problems.

Reply:
Actually, @bparees (or @pweil-) do either of you have an opinion about this kind of pattern about deleting objects we're currently processing in the worker queue? It looks like we're passing a deep copy that the worker grabbed from the cache, so we might be all set in terms of invalidating the cache, but I would imagine we should just mark the object as a deletion candidate, quit early, and delete that object when it's re-queued.

Reply:
i'm not positive i understand the scenario, but if this is a typical event/controller loop, i would expect you to not manipulate the informer cache directly, i would expect you to make a client api call to delete the resource itself(by name/namespace), which in turn will trigger an event back to your informer cache that will remove the entry from the cache (and let your controller loop receive an event that the resource is being deleted and handle that event as desired).

Reply:
The only thing I'd add to @bparees response is that I would argue that calling `deleteReport` from `runReport` is an unintended side effect and probably confusing.  Let `runReport` (and other methods) provide safety checks for bad input but if `runReport` actually runs, it only runs a report.  Otherwise it returns `nil` or errors but does not take further action.  Easier to test, easier to grok.

Let the orchestrator be responsible for state management and which method to call (`runReport`, `deleteReport`, `rescheduleReport`, etc) and let those method guard against bad input and execute.

Reply:
Moved to `syncReport` and used existing `op.deleteReport` instead of creating a new deleteReport. The existing `op.deleteReport` deletes the report from the queues and then, yep, as Ben said, it was needed to call the client delete to delete the Report resource.